File: baekjoon/3주차/2493-탑/제문/2493.kt
package baekjoon

import java.util.*

fun main() = with(System.`in`.bufferedReader()) {
    val towerNum = readLine().toInt()
    val tower = readLine().split(" ").map { it.toInt() }.toIntArray()
    val answer = IntArray(towerNum)
    val towersOnTheLeft = Stack<Pair<Int, Int>>() // Stack의 Top은 좌측의 타워 중 보이는 첫 번째 타워

    // answer[0]은 항상 0이므로 하드 코딩
    // Pair.first : index
    // Pair.second : tower height
    towersOnTheLeft.push(Pair(1, tower[0]))

    for (idx in 1 until towerNum) {
        while (towersOnTheLeft.isNotEmpty()) {
            val currentTower = towersOnTheLeft.peek()

            if (currentTower.second < tower[idx]) {
                towersOnTheLeft.pop()
            }
            else {
                answer[idx] = currentTower.first
                break
            }
        }
        towersOnTheLeft.push(Pair(idx + 1, tower[idx]))
    }

    print(answer.joinToString(" "))
}

File: baekjoon/3주차/2493-탑/정우/2493.py
import sys

sys.stdin.readline()
arr = list(map(int, sys.stdin.readline().split()))
ret = [0 for _ in range(len(arr))]
stack = [(len(arr)-1, arr[-1]),] # [(idx, value)]
for i in range(len(arr)-2, -1, -1):
    idx, value = stack[-1]
    while value < arr[i]:
        ret[idx] = i+1
        stack.pop()
        if stack:
            idx, value = stack[-1]
        else:
            break
    stack.append((i, arr[i]))
ret = list(map(str, ret))
print(' '.join(ret))
